FLT/NumberField/DiscriminantBounds.lean

```diff
@@ -0,0 +1,174 @@
+
+private lemma discrBound_strictMono_aux3 {x : ℝ} (hx : 1 ≤ x) : -0.5 ≤ log x / 2 - x / (x + 1) := by
+  suffices h : StrictMonoOn (fun x ↦ log x / 2 - x / (x + 1)) (Set.Ioi 0) by
+    convert h.monotoneOn (by simp) (by simp; linarith) hx using 1
+    norm_num
+  apply strictMonoOn_of_hasDerivWithinAt_pos (convex_Ioi _)
+    (f' := fun x ↦ (x ^ 2 + 1) / (2 * x * (x + 1) ^ 2))
+  case hf =>
+    refine ((continuousOn_log.mono (by simp)).div_const _).sub
+      (continuousOn_id.div (by fun_prop) ?_)
+    simp only [Set.mem_Ioi, ne_eq]
+    intro x hx
+    positivity
+  case hf'₀ =>
+    simp only [interior_Ioi, Set.mem_Ioi]
+    intro x hx
+    positivity
+  case hf' =>
+    simp only [interior_Ioi, Set.mem_Ioi]
+    intro x hx
+    refine .congr_deriv ((((hasDerivWithinAt_id _ _).log hx.ne').div_const _).sub
+      ((hasDerivWithinAt_id _ _).div ((hasDerivWithinAt_id _ _).add_const _) (by linarith))) ?_
+    field_simp
+    ring
+
+private lemma discrBound_strictMono_aux2 (n : ℕ) (hn : n ≠ 0) :
+    0 < n ^ 2 * (log (n + 1) - log n) + log n ! - n * log n := by
+  calc
+    _ > n ^ 2 * (log (n + 1) - log n) + 0.9 + log n / 2 - n := by
+        linear_combination Stirling.le_log_factorial_stirling n hn + constant_approx / 2
+    _ ≥ n ^ 2 * log (1 + (n : ℝ)⁻¹) + 0.9 + log n / 2 - n := by
+        rw [← log_div (by positivity) (by positivity)]; field_simp
+    _ ≥ n ^ 2 * (1 + n : ℝ)⁻¹ + 0.9 + log n / 2 - n := by
+        gcongr; exact log_approx (by positivity)
+    _ = log n / 2 - n / (n + 1) + 0.9 := by
+        field_simp; ring
+    _ ≥ 0 := by
+        linear_combination discrBound_strictMono_aux3 (x := n) (by simp; omega)
+
+-- upstream
+theorem strictMonoOn_Ici_nat_of_lt_succ {α : Type*} [Preorder α] {f : ℕ → α} {x : ℕ}
+    (h : ∀ n, x ≤ n → f n < f (n + 1)) :
+    StrictMonoOn f {i | x ≤ i} :=
+  fun _ ha _ _ ↦ Nat.rel_of_forall_rel_succ_of_le_of_lt (· < ·) h ha
+
+-- upstream
+theorem strictMonoOn_Ioi_nat_of_lt_succ {α : Type*} [Preorder α] {f : ℕ → α} {x : ℕ}
+    (h : ∀ n, x < n → f n < f (n + 1)) :
+    StrictMonoOn f (Set.Ioi x) :=
+  strictMonoOn_Ici_nat_of_lt_succ (x := x + 1) h
+
+private lemma discrBound_strictMono_aux :
+    StrictMonoOn (fun n : ℕ ↦ log n - (n : ℝ)⁻¹ * log n !) (Set.Ioi 0) :=
+  strictMonoOn_Ioi_nat_of_lt_succ fun n hn ↦ by
+    refine lt_of_mul_lt_mul_left (a := n * (n + 1 : ℝ)) ?_ (by positivity)
+    calc
+      _ = (n * (n + 1)) * log n - (n + 1) * log n ! := by field_simp; ring
+      _ < n * (n + 1) * log (n + 1) - n * (log (n + 1) + log n !) := by
+          linear_combination discrBound_strictMono_aux2 n hn.ne'
+      _ = n * (n + 1) * (log (n + 1) - (n + 1 : ℝ)⁻¹ * (log (n + 1) + log n !)) := by
+          field_simp; ring
+      _ = n * (n + 1) * (log (n + 1 : ℕ) - (↑(n + 1) : ℝ)⁻¹ * log (n + 1)!) := by
+          rw [Nat.factorial_succ, Nat.cast_mul, log_mul (by positivity) (by positivity),
+            Nat.cast_add_one]
+
+noncomputable def discrBound (n : ℕ) : ℝ :=
```

Comment:
```suggestion
/--
Minkowski's lower bound for the root discriminant (that is, the n'th root of the absolute value of the discriminant) of a totally complex number field of degree n.
-/
noncomputable def rootDiscrBound (n : ℕ) : ℝ :=
```

Reply:
I think it's better being called rootDiscrBound but other than that this looks good to go. Thanks!

Reply:
You'd have hated the name I gave it in the first commit then :D